### ISSUE TITLE
Seems like Tween.js won't animate getters/setters.

### ISSUE THREAD

**Issue by trusktr** [tweenjs/tween.js#267]
I don't know why, but in my case the object who's values I'm trying animate isn't being animated. I have code like this:

```js
let positionTween = new TWEEN.Tween(letterNode.position)
    .to(cellPosition, 1000)
    .onUpdate(function() {
        console.log('   final position', cellPositions)
        console.log('     tween value?', this)
        console.log('  actual position', letterNode.position)
    })
    .onComplete(function() { positionTween.completed = true })
    
let task = Motor.addRenderTask(time => {
    if (!positionTween.started) {
        positionTween.start(time)
        positionTween.started = true
    }
    else positionTween.update(time)
    
    if (positionTween.completed) {
        console.log('completed')
        Motor.removeRenderTask(Task)
    }
})
```

The problem is demonstrated by the console.log output.

![screen shot 2016-06-12 at 2 05 23 am](https://cloud.githubusercontent.com/assets/297678/15990232/2b23a3ea-3042-11e6-9ff7-47dca951e8b6.png)

In the screenshot we can see what the final position should be, and what the actual tween values are. For some reason, the actual values aren't changing, and are not approaching the expected end values.

Am I missing something?

**Comment by trusktr** [comment]
I added another console.log statement to show that the object I'm trying to animate is the same object as `this` inside of `onUpdate`:

```js
let positionTween = new TWEEN.Tween(letterNode.position)
    .to(cellPosition, 1000)
    .onUpdate(function() {
        console.log('   final position', cellPositions)
        console.log('     tween value?', this)
        console.log('  actual position', letterNode.position)
        console.log('     same object?', letterNode.position === this) // true!
    })
    .onComplete(function() { positionTween.completed = true })
    
let task = Motor.addRenderTask(time => {
    if (!positionTween.started) {
        positionTween.start(time)
        positionTween.started = true
    }
    else positionTween.update(time)
    
    if (positionTween.completed) {
        console.log('completed')
        Motor.removeRenderTask(Task)
    }
})
```

and the output shows `true`. So, that seems to be all good, but I don't understand why the values aren't changing:

![screen shot 2016-06-12 at 2 18 39 am](https://cloud.githubusercontent.com/assets/297678/15990304/fe53bb1e-3043-11e6-8c5e-99c3de1df2cc.png)

By the way, the `x`, `y`, and `z` properties that I'm trying to animate are `getters/setters`. For example, if I open one of the XYZValues objects, you can see it has the matching getters:

![screen shot 2016-06-12 at 2 19 51 am](https://cloud.githubusercontent.com/assets/297678/15990314/2d107e7e-3044-11e6-8297-eac74cf2ffce.png)

The setters/getters store values into the corresponding underscored properties.

**Comment by trusktr** [comment]
I was able to work around the issue: instead of providing the actual object to `new Tween`, I made a made a new object, then update the values in `onUpdate`, and it works! So I changed the code to this:

```js
let positionTween = new TWEEN.Tween({
    x: letterNode.position.x,
    y: letterNode.position.y,
    z: letterNode.position.z,
})
    .to(cellPosition, 1000)
    .onUpdate(function() {
        console.log('   final position', cellPositions)
        console.log('     tween value?', this)
        console.log('  actual position', letterNode.position)
        console.log('     same object?', letterNode.position === this) // true!
        
        letterNode.position.x = this.x
        letterNode.position.y = this.y
        letterNode.position.z = this.z
    })
    .onComplete(function() { positionTween.completed = true })
    
let task = Motor.addRenderTask(time => {
    if (!positionTween.started) {
        positionTween.start(time)
        positionTween.started = true
    }
    else positionTween.update(time)
    
    if (positionTween.completed) {
        console.log('completed')
        Motor.removeRenderTask(Task)
    }
})
```


Any idea why it works when I do it the second way, but not the first way?

**Comment by trusktr** [comment]
I've been pulling y hair out twiddling with my code for hours, until finally it worked when just using a new object instead of the target object directly. But, I'd like to use the target object directly because it's cleaner.

**Comment by mikebolt** [comment]
This is probably a bug in Tween.js. If you could make a codepen that reproduces the issue that would be a big help, because then I could just step through the update function.

If those are all the properties and methods of XYZValues, then I might be able to reproduce this on my own.

**Comment by mikebolt** [comment]
I was unable to reproduce this issue. Animating properties that have both a getter and setter method seems to work for me. Here is my attempt to reproduce the issue:

https://github.com/mikebolt/tween.js/blob/issue267/examples/002_get_and_set_closure.html

I will need the code for the XYZValues constructor to properly investigate this issue.

**Comment by sole** [comment]
Could it be a combination of trying to guess which properties are in the object itself, in https://github.com/tweenjs/tween.js/blob/master/src/Tween.js#L117-L119 and then values in the object at animation time in https://github.com/tweenjs/tween.js/blob/master/src/Tween.js#L160 ?

**Comment by trusktr** [comment]
@sole I ran a small test in Chrome console:

```js
let o = {
  set n(v) {this._n = v},
  get n() {return this._n}
}
o.n = 5
for (var i in o) console.log(i)
```

The for-in loop does indeed log the `n` getter/setter, so it does see it.

Also, none of my starting values are `undefined`. The object that I've passed into `TWEEN.Tween` has pre-defined numerical values set by default when constructed, so that check for `undefined` doesn't seem like it will throw anything off. 

I'm not sure what's happening. Maybe I can put a bunch of console.logs everywhere and see the difference in output with literal properties vs getters/setters...

**Comment by sole** [comment]
Thanks for looking into it.

When the library was first written there was no way of having setters and getters on JS :-)

**Comment by brenwell** [comment]
Hey everyone, I was have been using tween.js with pixi.js for a while now and have been having the same issues, I believe, with the latest version.

```coffeescript
toObj = {
      y: 800,  
      alpha: 0, 
      rotation: Math.random()
}
new TWEEN.Tween(snowflake).to(toObj,1400).start()  
```

The Y value is nolonger interpolated via the setter. I have to split it to get it to work. Not sure if its an intentional change.

```coffeescript
toObj = {
      alpha: 0, 
      rotation: Math.random()
}
new TWEEN.Tween(snowflake).to(toObj,1400).start()  

toPosObj = {
      y: 800
}
new TWEEN.Tween(snowflake.position).to(toPosObj,1400).start() 
```

Just throwing my 2 cents in here. Thanks

**Comment by mikebolt** [comment]
Maybe the problem is caused by transpiling the code? To emulate getters and
setters it would have to replace certain assignment expressions with
function calls, but it would not be able to replace e.g. `obj[propertyName]
= 4` with its appropriate setter unless it knew the value of `propertyName`.

**Comment by mikebolt** [comment]
I meant to ask if you were transpiling the code. If so, can you post the
unminified transpiled version?
On Jun 22, 2016 10:21 AM, "Michael Casebolt" <mikebolt@gmail.com> wrote:

Maybe the problem is caused by transpiling the code? To emulate getters and
setters it would have to replace certain assignment expressions with
function calls, but it would not be able to replace e.g. `obj[propertyName]
= 4` with its appropriate setter unless it knew the value of `propertyName`.

**Comment by trusktr** [comment]
I'm transpiling with a recent version of Babel. Setters/getters [aren't transpiled by default](https://babeljs.io/repl/#?evaluate=true&lineWrap=false&presets=es2015%2Creact&code=let%20o%20%3D%20%7B%0A%20%20set%20n(v)%20%7Bthis._n%20%3D%20v%7D%2C%0A%20%20get%20n()%20%7Breturn%20this._n%7D%2C%0A%7D) (I'm using the same settings).

It seems to be a runtime issue.

**Comment by trusktr** [comment]
@brenwell In your case is a different issue: there's no way the `y` value in the `toObj` can map to `position.y` of the `snowflake`. Tween.js looks for the matching property names on the object passed to the Tween constructor, and `snowflake.y` doesn't exist, which is what Tween.js looks for based on the `toObj`. That's why it works when you pass the `snowflake.position` object to the Tween constructor.

**Comment by martikaljuve** [comment]
One solution is to copy only those starting values from the target object that have destination values in `.to(properties, duration)`, as can be seen in https://github.com/martikaljuve/tween.js/commit/a50474b416b73554559e882ff3e22fe056c75078. Unit tests are passing, but I haven't considered if/how this would break other use cases.

**Comment by dalisoft** [comment]
@trukstr Try to enable `{writable:1,enumerable:1}` to `setter` to allow change and tween (tween only numbers)

**Comment by mikebolt** [comment]
@martikaljuve That fix looks good to me. The behavior would change: the tweened object's starting properties would be saved when `.to` is called instead of when the tween is constructed. We can bump the major version number in npm to indicate that there is a minor backwards compatibility issue. I don't think it's a major break because we never specified in the documentation exactly when the starting properties are saved.

We need some new unit tests for tweening getter/setter properties before I can pull the fix and close this issue.

**Comment by trusktr** [comment]
@martikaljuve 

> Using a class or Object.defineProperty creates non-enumerable properties:

That seems horrible. 

@dalisoft 

> Try to enable {writable:1,enumerable:1} to setter to allow change and tween (tween only numbers)

It works! But the solution is ugly when using classes. For example, this is what one of my classes looks like now:

```js
/**
 * Represents a set of values for the X, Y, and Z axes. For example, the
 * position of an object can be described using a set of 3 numbers, one for each
 * axis in 3D space: {x:10, y:10, z:10}.
 *
 * The value don't have to be numerical. For example,
 * {x:'foo', y:'bar', z:'baz'}
 */
class XYZValues {
    constructor(x = 0, y = 0, z = 0) {
        this._x = x
        this._y = y
        this._z = z
    }

    // override this on the instance to run logic on a property change.
    onChanged() {}

    set x(value) {
        this._x = value
        this.onChanged()
    }
    get x() { return this._x }

    set y(value) {
        this._y = value
        this.onChanged()
    }
    get y() { return this._y }

    set z(value) {
        this._z = value
        this.onChanged()
    }
    get z() { return this._z }
}

const xDesc = Object.getOwnPropertyDescriptor(XYZValues.prototype, 'x')
xDesc.enumerable = true
Object.defineProperty(XYZValues.prototype, 'x', xDesc)

const yDesc = Object.getOwnPropertyDescriptor(XYZValues.prototype, 'y')
yDesc.enumerable = true
Object.defineProperty(XYZValues.prototype, 'y', yDesc)

const zDesc = Object.getOwnPropertyDescriptor(XYZValues.prototype, 'z')
zDesc.enumerable = true
Object.defineProperty(XYZValues.prototype, 'z', zDesc)

export {XYZValues as default}
```

**Comment by trusktr** [comment]
I made a small helper, which I guess I'll be using all the time:

```js
function makeAccessorsEnumerable(object) {
    const props = Object.getOwnPropertyNames(object)
    for (let prop of props) {
        const descriptor = Object.getOwnPropertyDescriptor(object, prop)
        if (descriptor && (descriptor.set || descriptor.get)) {
            descriptor.enumerable = true
            Object.defineProperty(object, prop, descriptor)
        }
    }
}
```

then the code is cleaner:

```js
import {makeAccessorsEnumerable} from './Utility'

class XYZValues {
    constructor(x = 0, y = 0, z = 0) {
        this._x = x
        this._y = y
        this._z = z
    }

    onChanged() {}

    set x(value) {
        this._x = value
        this.onChanged()
    }
    get x() { return this._x }

    set y(value) {
        this._y = value
        this.onChanged()
    }
    get y() { return this._y }

    set z(value) {
        this._z = value
        this.onChanged()
    }
    get z() { return this._z }
}

// This is needed in order for Tween.js to detect and use the accessors,
// otherwise Tween.js won't be able to tween them.
makeAccessorsEnumerable(XYZValues.prototype)

export {XYZValues as default}
```

**Comment by dalisoft** [comment]
Whoops, i am not a advanced developer. My mine limited. I am not think even about Classes, because don't know about it. I hope you find way to work it

**Comment by martikaljuve** [comment]
@trusktr, you could also use ES7 decorators:

```js
// utils/decorators.js
export function enumerable(target, key, descriptor) {
	descriptor.enumerable = true;
	return descriptor;
}
```
```js
// xyz_values.js
import { enumerable } from './utils/decorators';

export default class XYZValues {
	constructor(x = 0, y = 0, z = 0) {
		this._x = x;
		this._y = y;
		this._z = z;
	}

	onChanged() {}

	@enumerable
	get x() { return this._x; }
	set x(value) { this._x = value; this.onChanged(); }

	@enumerable
	get y() { return this._y; }
	set y(value) { this._y = value; this.onChanged(); }

	@enumerable
	get z() { return this._z; }
	set z(value) { this._z = value; this.onChanged(); }
}
```

**Comment by trusktr** [comment]
@martikaljuve Decorators would be nice! But seems the spec is being changed, and Babel dropped support for now, so will wait on that. I wanted to use decorators for some other things too, but just waiting.

**Comment by trusktr** [comment]
@martikaljuve In your example, would

```js
    @enumerable
    get z() { return this._z; }
    set z(value) { this._z = value; this.onChanged(); }
```

be equivalent to

```js
    get z() { return this._z; }
    @enumerable
    set z(value) { this._z = value; this.onChanged(); }
```

or

```js
    @enumerable
    get z() { return this._z; }
    @enumerable
    set z(value) { this._z = value; this.onChanged(); }
```

?

**Comment by trusktr** [comment]
@sole For the documentation, maybe you can just make clear that Tween.js works with *enumerable* properties, and briefly describe that with getter/setters on classes we have to modify the descriptors so they are enumerable. You could show my makeAccessorsEnumerable helper, if that would help.

**Comment by martikaljuve** [comment]
@trusktr, the decorator most likely works the same whether it's on the getter or the setter. `babel-plugin-transform-decorators-legacy` transpiles it to `Object.getOwnPropertyDescriptor(XYZValues.prototype, 'z')` in both cases.

**Comment by sole** [comment]
Thanks for the investigations!

@trusktr: what do you think about making a PR with these suggestions, as you can express your findings more easily than me? That would be lovely

**Comment by simonmurdock** [comment]
Ran into this issue also, Using a transpired Aurelia application, and when trying to tween properties on a singleton class injected into my class it won't work. Spent an hour wondering WTF is going on.

Worked when I used an intermediate object like @trusktr suggested.

**Comment by mikebolt** [comment]
This issue should be fixed now with jonobr's PR #291. @martikaljuve's fix would have worked great too. In retrospect I should have merged it, and neither jonobr nor simonmurdock would have ran into this issue. I hope you don't blame me for being confused :)

As a side note, if you are using the old version of the code and defining properties with Object.defineProperty, you can make the properties enumerable by passing "true" as the second parameter. I think this is the workaround that jonobr used.

**Comment by trusktr** [comment]
Cool! I don't need to make my class accessors enumerable now, and @simonmurdock you don't need your workaround anymore.
